circuits/cpp/src/aztec3/utils/array.hpp

```diff
@@ -43,6 +43,23 @@ template <size_t SIZE> NT::fr array_length(std::array<NT::fr, SIZE> const& arr)
     return length;
 };
 
+/**
+ * Gets the number of contiguous elements until we encounter one that reports itself as 'empty'.
+ * Note: This assumes `0` always means 'not used', so be careful. As soon as we locate 0, we stop the counting.
+ * If you actually want `0` to be counted, you'll need something else.
+ */
+template <typename T, size_t SIZE> size_t array_length(std::array<T, SIZE> const& arr)
+{
+    size_t length = 0;
+    for (const auto& e : arr) {
+        if (e.is_empty()) {
```

Comment:
I created a spike issue to explore how we could improve the code to make use of 'concepts'. https://github.com/AztecProtocol/aztec3-packages/issues/343
(Note: this trend of assuming `is_empty` existed before this PR, so no need to modify this function for this PR)